### PR TITLE
jgcastanos2-patch1-issue51

### PR DESCRIPTION
**Pull request by jcastanos2**
First version of how to support both updates of the bmc and the cec in NVIDIA Bluefields.
- get_fw_facts now returns a dictionary of facts, rather than individual get_xxxx facts
- in bmc_fw_update, no need to get original fw (not used anywhere?), and passing the target fw as a string, rather than using filename. There's no real mapping between filenames and versions returned by the bmc
- allow call multiple times in meta/main.yml
- a sample update_fw.yml at the top shows how the bmc_fw_update can be called twice, and which vars to pass in each instance. There's a sample cmd line invocation at the top
- this file also shows how to check the update was successful after doing a power cycle of the server. Checks immediately after the update might be meaningless. 
- This update_fw.yml is Bluefield specific. It can be removed, but I am not sure how to document/explain the whole flow to the customers without it. It can also be adapted to fit customers needs. I see it more as part of documentation on how to use, rather than a strong role dependency.

### CODE REVIEW

File: roles/bmc_fw_update/tasks/main.yml [hunk 1]
```diff
@@ -76,18 +76,22 @@
   when:
     - bmc_fw_update_reboot is true
 
+- name: Get the DPU a chance to restart
+  ansible.builtin.pause:
+    seconds: 120
```
Comment (glimchb) [inline roles/bmc_fw_update/tasks/main.yml:81]: this is already in the reboot role

File: roles/bmc_fw_update/tasks/main.yml [hunk 2]
```diff
@@ -4,14 +4,14 @@
 ---
 # tasks file for bmc_fw_update
 
-- name: Get Firmware Inventory
-  ansible.builtin.include_role:
-    name: get_bmc_facts
-  register: get_bmc_facts_before
```
Comment (glimchb) [inline roles/bmc_fw_update/tasks/main.yml:10]: why this is commented out ?

File: roles/get_bmc_facts/tasks/main.yml [hunk 3]
```diff
@@ -16,5 +16,10 @@
 
 - name: Extract all FW versions to facts
   ansible.builtin.set_fact:
```
Comment (glimchb) [inline roles/get_bmc_facts/tasks/main.yml:18]: this looks like a separate change we want to merge first, can you do separate PR for i t?

File: roles/get_bmc_facts/tasks/main.yml [hunk 4]
```diff
@@ -16,5 +16,10 @@
...
-    "{{ 'get_bmc_facts_' + item['Id'] | lower }}": "{{ item['Version'] }}"
+    bmc_fw_version: "{{ bmc_fw_version | default({}) | combine({ item['Id'] | lower : item['Version'] }) }}"
   loop: "{{ result.redfish_facts.firmware.entries }}"
```
Comment (glimchb) [inline roles/get_bmc_facts/tasks/main.yml:20]: `bmc_fw_version` looks like wrong var name... we have here all the inventory, no ?

File: update_bf.yml [hunk 5]
```diff
@@ -0,0 +1,62 @@
+# exmaple run to update both the bmc fw and the cec in an NVIDIA BlueField
+# ansible-playbook -i ../inventory/bfbmcs.ini --extra-vars "dpu_bmc_username=${BMCUSER} dpu_bmc_password=${BMCPW} bmc_fw_filename=/home/sancho/bmcimages/bf3-bmc-23.10-5_opn.fwpkg bmc_cec_filename=/home/sancho/bmcimages/cec1736-ecfw-00.02.0152.0000-n02-rel-prod.fwpkg fw_version=BF-23.10-5 cec_version=00.02.0152.0000_n02" update_bf.yml
```
Comment (glimchb) [inline update_bf.yml:2]: file belongs in the `playbook` folder ?

**Review by glimchb — CHANGES_REQUESTED** [review]
thanks
few comments inside

File: roles/bmc_fw_update/tasks/main.yml (hunk 2 above)
Comment (jcastanos2) [reply roles/bmc_fw_update/tasks/main.yml:10]: why do we need it? It is not being used. And when we update the bmc and the cec, the second time we get a meaningless get_bmc_facts_before (after the bmc and before the cec ...?)

File: roles/get_bmc_facts/tasks/main.yml (hunk 4 above)
Comment (jcastanos2) [reply roles/get_bmc_facts/tasks/main.yml:20]: I agree it is a beyond confusing variable name ... I was using version as in "Version" from redfish ... Fact also has a bigger name in ansible; this is just a part of the facts ... If we cannot come up with something better, I'd be happy to move it back to "fact"

File: roles/bmc_fw_update/tasks/main.yml (hunk 2 above)
Comment (glimchb) [reply roles/bmc_fw_update/tasks/main.yml:10]: if we don't need it, remove it :-) don't keep comment out code

File: roles/get_bmc_facts/tasks/main.yml (hunk 4 above)
Comment (glimchb) [reply roles/get_bmc_facts/tasks/main.yml:20]: how about `get_bmc_facts_all_fw_versions` ?
